lib/ruby_smb/dcerpc/ndr.rb

```diff
@@ -92,6 +95,85 @@ def set(v)
         end
       end
 
+      # An NDR Uni-dimensional Fixed Array of bytes representation as defined in:
+      # [Transfer Syntax NDR - NDR Constructed Types](https://pubs.opengroup.org/onlinepubs/9629399/chap14.htm#tagcjh_19_03_03_01)
+      class NdrFixedByteArray < BinData::BasePrimitive
+        optional_parameters :length, :pad_byte, :pad_front
+        default_parameters pad_byte: 0
+        mutually_exclusive_parameters :length, :value
+
+        def assign(val)
+          super(fixed_byte_array(val))
+        end
+
+        def snapshot
+          clamp_to_length(super)
+        end
+
+        class << self
+          def arg_processor
+            NdrFixedByteArrayArgProcessor.new
+          end
+        end
+
+        private
+
+        def clamp_to_length(val)
+          val = fixed_byte_array(val)
+          len = eval_parameter(:length) || val.length
+          if val.length > len
+            val = val.first(len)
+          elsif val.length < len
+            pad = eval_parameter(:pad_byte)
+            if get_parameter(:pad_front)
+              val = val.insert(0, *Array.new(len - val.length, pad))
+            else
+              val = val.fill(pad, val.length...len)
+            end
+          end
+
+          val
+        end
+
+        def fixed_byte_array(val)
+          val = val.bytes if val.is_a? String
+          val.to_ary
+        end
+
+        def read_and_return_value(io)
+          len = eval_parameter(:length) || 0
```

Comment:
If `:length` parameter is not provided, nothing will be read. This can be confusing for the user since this parameter is not mandatory. Maybe `:length` could be set as a `mandatory_parameter`?

Reply:
That makes sense. In commit 013c9ed1ae654f96a1a1f7dc8bc24ffb06c0dc0b I implemented a `WarnNoReadLengthPlugin` so this behaves just like the `BinData::String` [class](https://github.com/dmendel/bindata/blob/5a523e2d826763a7505d73920de1a09ffdb4ef23/lib/bindata/string.rb) from which it is derived. When a value is specified without a `:read_length` a warning will be thrown. If neither is specified, then it's treated as an empty array just like the related string class is treated as an empty string.

Reply:
Great! Sounds good to me! Thanks!